src/mlpack/methods/kde/kde_rules_impl.hpp

```diff
@@ -0,0 +1,130 @@
+/**
+ * @file kde_rules_impl.hpp
+ * @author Roberto Hueso (robertohueso96@gmail.com)
+ *
+ * Implementation of rules for Kernel Density Estimation with generic trees.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_METHODS_KDE_RULES_IMPL_HPP
+#define MLPACK_METHODS_KDE_RULES_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "kde_rules.hpp"
+
+namespace mlpack {
+namespace kde {
+
+template<typename MetricType, typename KernelType, typename TreeType>
+KDERules<MetricType, KernelType, TreeType>::KDERules(
+    const arma::mat& referenceSet,
+    const arma::mat& querySet,
+    arma::vec& densities,
+    const double error,
+    const std::vector<size_t>& oldFromNewQueries,
+    MetricType& metric,
+    const KernelType& kernel) :
+    referenceSet(referenceSet),
+    querySet(querySet),
+    densities(densities),
+    error(error),
+    oldFromNewQueries(oldFromNewQueries),
+    metric(metric),
+    kernel(kernel),
+    lastQueryIndex(querySet.n_cols),
+    lastReferenceIndex(referenceSet.n_cols),
+    baseCases(0),
+    scores(0)
+{
+  // Nothing to do.
+}
+
+//! The base case.
+template<typename MetricType, typename KernelType, typename TreeType>
+inline force_inline
+double KDERules<MetricType, KernelType, TreeType>::BaseCase(
+    const size_t queryIndex,
+    const size_t referenceIndex)
+{
+  double distance = metric.Evaluate(querySet.col(queryIndex),
+                                    referenceSet.col(referenceIndex));
+  densities(oldFromNewQueries.at(queryIndex)) += kernel.Evaluate(distance);
+  ++baseCases;
+  lastQueryIndex = queryIndex;
+  lastReferenceIndex = referenceIndex;
+  return distance;
+}
+
+//! Single-tree scoring function.
+template<typename MetricType, typename KernelType, typename TreeType>
+double KDERules<MetricType, KernelType, TreeType>::
+Score(const size_t /* queryIndex */, TreeType& /* referenceNode */)
+{
+  ++scores;
+  traversalInfo.LastScore() = 0.0;
+  return 0.0;
+}
+
+template<typename MetricType, typename KernelType, typename TreeType>
+double KDERules<MetricType, KernelType, TreeType>::Rescore(
+    const size_t /* queryIndex */,
+    TreeType& /* referenceNode */,
+    const double oldScore) const
+{
+  // If it's pruned it continues to be pruned.
+  return oldScore;
+}
+
+//! Double-tree scoring function.
+template<typename MetricType, typename KernelType, typename TreeType>
+double KDERules<MetricType, KernelType, TreeType>::
+Score(TreeType& queryNode, TreeType& referenceNode)
+{
+  double score, bound;
+  bound = kernel.Evaluate(queryNode.MinDistance(referenceNode)) -
+          kernel.Evaluate(queryNode.MaxDistance(referenceNode));
+
+  if (bound <= error / referenceSet.n_cols)
```

Comment:
I think this is absolute error pruning.  I guess we can use both (absolute and relative), and maybe it is easy to give the user the option of which to choose.  I think the "Tree-Independent Dual-Tree Algorithms" paper has relative-value approximation pruning rules---if they are not there let me know, I can dig them up or otherwise write them.  (It is not too hard a change.)

Reply:
I think it just has an absolute error approach.
What do you think about something like this?

```C++
double score, bound, min;
bound = kernel.Evaluate(queryNode.MinDistance(referenceNode)) -
        kernel.Evaluate(queryNode.MaxDistance(referenceNode));
min_dist = kernel.Evaluate(queryNode.MinDistance(referenceNode));
if (bound <= (abs_error + rel_error * min_dist) / referenceSet.n_cols)
{...
```

Reply:
I think it would be very odd for someone to use a combined absolute error and relative error approach, but (assuming that prune is correct, and I *think* it is but I have not double-checked) if you are planning to force via warnings or errors that `abs_error = 0` when `rel_error > 0` and `rel_error = 0` when `abs_error > 0` I think that's fine.